### ISSUE TITLE
[UUID] Bug with Ramsey\UUID uuid_binary Doctrine type

### ISSUE THREAD

**Issue by dkarlovi** [api-platform/core#1309]
If using Ramsey Doctrine UUID `uuid_binary` datatype, identifiers in motion are classic UUID strings (ie. `5e619188-b5c8-4797-905e-05f4ba6dfc67`) but they're binary at rest (ie. `0xFEFDF85691AE48E2888D39C3EBEC352F`).

[In IdentifierManagerTrait, a statement tries to convert the user-supplied identifier to a PHP value](https://github.com/api-platform/core/blob/master/src/Bridge/Doctrine/Orm/Util/IdentifierManagerTrait.php#L80), which means it tries to convert the input as if it's binary, NOT string. This fails, obviously. If I change it to `convertToDatabaseValue()`, it works as expected.

This brings up another question, if the ID provided is not decodable (not a valid UUID, for example), how should that be treated? I guess as a 404, not as a 500.

/cc @teohhanhui 

TLDR
```
GET http://127.0.0.1:8000/api/images/5e619188-b5c8-4797-905e-05f4ba6dfc67

{
    "type": "https://tools.ietf.org/html/rfc2616#section-10",
    "title": "An error occurred",
    "detail": "Could not convert database value \"5e619188-b5c8-4797-9...\" to Doctrine Type uuid_binary"
}
```

**Comment by theofidry** [comment]
>if the ID provided is not decodable (not a valid UUID, for example), how should that be treated? I guess as a 404, not as a 500

For a URL, if the ID is not a valid UUID or the entity of that ID doesn't exists it should be a 404.

**Comment by meyerbaptiste** [comment]
I'm not familiar with this part of the core but this bug makes sense to me.

`Type::convertToPHPValue()` method allows Doctrine to normalize the data in PHP from the output of the database, therefore it expects in our case to receive many bytes. 
So, IMO, we don't use properly this method because we should not use it to normalize data from a client.

@api-platform/core-team why don't we use the serializer to normalize IDs here?

**Comment by dkarlovi** [comment]
@meyerbaptiste a proper way to fix this might be to use the serializer, yes. Especially since [the instance supports it](https://github.com/ramsey/uuid/blob/master/src/UuidInterface.php#L24).

**Comment by soyuka** [comment]
> @api-platform/core-team why don't we use the serializer to denormalize IDs here? For example in our case, this could allow us to write an UuidBinaryNormalizer to transform our stringified uuid to an Uuid instance. But maybe I'm totally wrong (because I admit it, it's weird to denormalize a value that comes from the URI)...

In fact, this is called from the `ItemDataProvider` which is linked to Doctrine. Indeed, it checks for the Doctrine identifiers to match them with the query parameter. 

The type transformation was added [when fixing types](https://github.com/api-platform/core/pull/1014/files).  Though I don't see a reason to transform an id to a Doctrine Type, it may be better to denormalize it instead? Also, you've a point where the data doesn't come from the database.

We're sending raw query attributes from the [ReadListener](https://github.com/api-platform/core/blob/183526946637ecf72ce1ddc4032ede281ee3b381/src/EventListener/ReadListener.php#L103).
Maybe we should denormalize them before they hit DataProviders?

**Comment by teohhanhui** [comment]
`convertToPHPValue` is indeed what we want - the identifier in the correct PHP type.

Can you help us understand why the conversion is failing?

**Comment by dkarlovi** [comment]
@teohhanhui `convertToPHPValue` expects the input to be in database format which it isn't if's coming from the user.

**Comment by teohhanhui** [comment]
> We're sending a raw query attribute from the ReadListener.
Maybe we should denormalize it before hitting DataProviders?

That sounds reasonable. But I don't know if we have enough information to do it at that layer...

**Comment by dkarlovi** [comment]
BTW this is tagged as a question, but it's a bug.

**Comment by dkarlovi** [comment]
Does anyone have an idea how to move forward here? I might send a PR to fix it if it's not too complex.

**Comment by soyuka** [comment]
@dkarlovi as I see it we need some kind of URLIdentiferDenormalization of the URL query parameters before they reach the ReadListener (see my comment above). Meaning that identifiers should be transformed to their correct type (in this case uuid, but there's also composite identifier, numbers, dates - can a date be an identifier? etc.).

**Comment by dkarlovi** [comment]
@soyuka I see, so would it use the serializer or have custom strategies? Or both?

**Comment by soyuka** [comment]
I don't think we can use the serializer it's not meant for that. I'm just using the same wording because it's currently `src/Bridge/Doctrine/Orm/Util/IdentifierManagerTrait.php`. Still, not sure if it's the best thing that this is related to Doctrine. 

Maybe we should do what we can before it reaches our data providers (https://github.com/api-platform/core/blob/183526946637ecf72ce1ddc4032ede281ee3b381/src/EventListener/ReadListener.php#L103).

This also reminds me of https://github.com/api-platform/core/issues/1376#issuecomment-330137535 (still not sure how I can put this).

**Comment by soyuka** [comment]
```php
class CompositeIdentifierNormalizer
{
    const COMPOSITE_IDENTIFIER_REGEX = '/(\w+)=(?<=\w=)(.+?)(?=;\w+=)|(\w+)=(.+)$/';

    public function normalize($data, $format = null, array $context = [])
    {
        $matches = [];
        $identifiers = [];
        $num = preg_match_all(self::COMPOSITE_IDENTIFIER_REGEX, $data, $matches, PREG_SET_ORDER);

        foreach ($matches as $i => $match) {
            if ($i === $num - 1) {
                $identifiers[$match[3]] = $match[4];
            } else {
                $identifiers[$match[1]] = $match[2];
            }
        }

        return $identifiers;
    }

    public function supportsNormalization($data, $format = null, array $context = []) {
        return $data && is_string($data) && preg_match_all(self::COMPOSITE_IDENTIFIER_REGEX, $data) > 0;
    }

    public function denormalize($data, $type, $format, array $context = []) {
        $identifier = '';

        foreach($data as $key => $value) {
            if ($identifier) {
                $identifier .= ";";
            }

            $identifier .= "$key=>$value";
        }

        return $identifier;
    }

    public function supportsDenormalization($data, $format = null, array $context = []) {
        return $data && is_array($data);
    }
}

// Test data
//
// $id = new IdentifierNormalizer();
//
// $id->normalize('a=bd;dc=d');
// $id->normalize('a=b;c=d foo;d=e');
// $id->normalize('a=1;c=2;d=10-30-24');
// $id->normalize('a=test;b=bar;foo;c=123');
// $id->normalize('a=test;b=bar ;foo;c=123;459;barz=123asgfjasdg4');
```

**Comment by dkarlovi** [comment]
@soyuka I like your "identifier normalizer" approach, but would like to have it be pluggable. For me, it would mean converting UUIDs as mentioned before, for someone else it might be something else.

Also, as @theofidry mentioned, if the normalizer isn't able to transform the identifier, by default it should be treated as "Not found", not an "Invalid argument" (this might differ based on context, though, need to work it out).

A new service which you can decorate?

**Comment by soyuka** [comment]
Yes coming with a PR explaining stuff this was just to keep a backup :laughing:

**Comment by rogamoore** [comment]
I just stumbled on the same problem as described in the initial post.
Is there currently any workaround?

**Comment by soyuka** [comment]
For me to finish the PR 😄

**Comment by Perf** [comment]
Having the same issue, but with `uuid_binary_ordered_time`.
All IDs in tables are of type `uuid_binary_ordered_time`. When sending request like:
`http://domain.tld/api/users/aed80460-1fdd-11e8-bae3-0800273b4cc5`
getting a response with `ConversionException` raised in method `convertToPHPValue`:
`Could not convert database value "aed80460-1fdd-11e8-b..." to Doctrine Type uuid_binary_ordered_time`.

However if I will try to fetch the same entity from code, like:
```php
$user = $em->getRepository(User::class)->findOneBy(['id' => 'aed80460-1fdd-11e8-bae3-0800273b4cc5']);
dump($user);
```
it works without any issues.

Are there any workaround for this issue?

**Comment by soyuka** [comment]
@Perf works just fine with https://github.com/api-platform/core/pull/1478 we will merge it eventually :P

**Comment by Perf** [comment]
@soyuka thank you! would be nice to merge it to master asap, or to some dev branch, so it will be easy to get it via composer update ;)

**Comment by dkarlovi** [comment]
@Perf you can always fetch his branch directly with Composer.

**Comment by soyuka** [comment]
Best is to use my master branch it's an uptodate fork of master with this patch!

```
    "require": {
        "api-platform/core": "dev-master",
    },
    "repositories": [
      {
        "type": "vcs",
        "url": "http://github.com/soyuka/core"
      }
    ],

```

**Comment by Perf** [comment]
@dkarlovi  yep, true!
@soyuka  thank you!

**Comment by Simperfit** [comment]
The PR has been merged, this can be closed ;).

**Comment by soyuka** [comment]
Please stop using my fork also :)

**Comment by dkarlovi** [comment]
@soyuka amazing to get this merged, looked like @dunglas doesn't like you very much in that PR :D

**Comment by soyuka** [comment]
> @soyuka amazing to get this merged, looked like @dunglas doesn't like you very much in that PR :D

:laughing: why you say that, it's a really sensitive part of the code therefore it needed to be reviewed a lot. Also I've simplified it a lot thanks to the reviewers :).

**Comment by dkarlovi** [comment]
JK, I think @dunglas was a great gatekeeper for this feature.

Thank you for sticking with this. <3

**Comment by michaelfeinbier** [comment]
Hey there,
Im using the `uuid_binary_ordered_time` Type like @Perf does and I still encounter the problem in the current 2.2.6 release:

```
[critical] Uncaught PHP Exception Doctrine\DBAL\Types\ConversionException: "Could not convert database value "22589e3e-68a6-11e8-9..." to Doctrine Type uuid_binary_ordered_time" at
```

But only in the `GET`-Item Action like: 
`GET http://127.0.0.1:8000/api/years/22589e3e-68a6-11e8-91f7-a45e60d266a1`
The List-Action works fine and correctly serializes the ID:
```
GET http://127.0.0.1:8000/api/years
{
    "@context": "/api/contexts/Year",
    "@id": "/api/years",
    "@type": "hydra:Collection",
    "hydra:member": [
        {
            "@id": "/api/years/22589e3e-68a6-11e8-91f7-a45e60d266a1",
            "@type": "Year",
            "id": "22589e3e-68a6-11e8-91f7-a45e60d266a1",
            "year": 2018
        },
        {
            "@id": "/api/years/cbc289b8-68c3-11e8-a81a-a45e60d266a1",
            "@type": "Year",
            "id": "cbc289b8-68c3-11e8-a81a-a45e60d266a1",
            "year": 2015
        }
    ],
    "hydra:totalItems": 2
}
```

Did I miss something or is it not related to api-core?

**Comment by Perf** [comment]
Hi @michaelfeinbier 
yes, it's still not in 2.2.6 release, use `"api-platform/core": "dev-master"` - it's working

**Comment by redecs** [comment]
I see this is not included in 2.2.7 either. Any chance it will be included in 2.2.8? I don't feel that confident using dev-master. Thanks!

**Comment by antograssiot** [comment]
master is for 2.3 so it will not land in any 2.2.x version.

**Comment by soyuka** [comment]
> I see this is not included in 2.2.7 either. Any chance it will be included in 2.2.8? I don't feel that confident using dev-master. Thanks!

@dunglas will release a beta for 2.3 soon!
